### ISSUE TITLE
[TS2741] Explicitly typed destructuring from broader type fails

### ISSUE THREAD

**Issue by MKRhere** [microsoft/TypeScript#32987]
**TypeScript version:**

`Version 3.7.0-dev.20190817`

**Search Terms:**

* "TS2741"
* "Property is missing in type but required in type"
* TypeScript explicitly typed destructuring assignment error

**Code**

```ts
// The contents of these don't matter, only the types (assume they already exist)
const x: { [k: string]: string } = {}; // Basically a string dict
const a: { p: string; q: string } = { p: "", q: "" };

const { y } = x;
// y is inferred correctly as string

const { p }: { p: string } = a;
// subtyped destructuring with explicit type works

const { z }: { z: string } = x;
// subtyped destructuring with explicit type doesn't work, even though z should already inferred as string
```

**Motivation:**

I validate the type in runtime and assume the incoming type to be too broad (which it is) and subtype on assignment so TypeScript knows the correct types. This is a special case that fails for me because of this specific type of `x`.

**Expected behavior:**

No error on either case.

**Actual behavior:**

`error TS2741: Property 'z' is missing in type '{ [k: string]: string; }' but required in type '{ z: string; }'.`

**Playground Link:**

[TypeScript playground](http://www.typescriptlang.org/play/#code/PTAEBUAsFNQYwPYDsAu1UGdQIGahTBrACbIDkKoAtgIYpoBOANNkgDYCe+M+HADtAwAoREgyUAHgC5QAb1ABtANYzxDAJZIA5gF1VKDdtABfUAF45xgNwjk40DRny++w1qugAjq81aT5uVAXUAAiEJZvUJCTG1sxSnkuUwsJGxBQLnUsTRxoBgZoYngEfOg4FE4HLDVfITj7ZxMnIJ8jZIc0sAwAVwAjFH5C0GJBA27y7rdQAHcSpWF6hNAALybA5da-dtShdJ7+waKRtXGUSd9hhEEkChm5lmgAN3RuBG6tSAzQLIc2AppiJkkLkCgwijRqgZfEA)

**Related Issues:**

I couldn't find any.

**Comment by rusev** [comment]
With `{ z }:` you type the whole think, not just the `z`. So the type should be `const { z }: typeof x = x;` or `const { z }: { [xxx: string]: string } = x;`.

[Updated example](https://www.typescriptlang.org/play/index.html#code/PTAEBUAsFNQYwPYDsAu1UGdQIGahTBrACbIDkKoAtgIYpoBOANNkgDYCe+M+HADtAwAoREgyUAHgC5QAb1ABtANYzxDAJZIA5gF1VKDdtABfUAF45xgNwjk40DRny++w1qugAjq81aT5uVAXUAAiEJZvUJCTG1sxSnkuUwsJGxBQLnUsTRxoBgZoYngEfOg4FE4HLDVfITj7ZxMnIJ8jZIc0sAwAVwAjFH5C0GJBA27y7rdQAHcSpWF6hNAALybAhQlN1t1t-xSbUQaVgEY1gYFcUAkA1KF0nv7BopG1cZRJ32GEQSQKGbmWNAAG7obgIbpaSAZUBZBxsAo0YiZJC5AoMIo0aoGXxAA)

**Comment by MKRhere** [comment]
@rusev Consider this example:

```ts
const { p }: { p: string } = a;
```

`{ p: string }` is only a subtype of `{ p: string; q: string }`, but it works correctly.

**Comment by fatcerberus** [comment]
`{ p: string }` is actually a *supertype* of `{ p: string, q: string }`, not a subtype.

**Comment by RyanCavanaugh** [comment]
There's nothing we can do here that doesn't introduce an inconsistency.

Index signatures assume that if you index in with a certain key, then you pre-validated that key to be in the object and will get back a non-undefined value:
```ts
declare const map: { [s: string]: string };
const A: string = map["foo"]; // OK
```

However, index signatures *don't* assume that they magically satisfy any object:
```ts
declare const map: { [s: string]: string };
const B: { a: string, b: string, zz: string } = map; // Not OK
```

If you write
```ts
declare const map: { [s: string]: string };
const { a } = map;
```
this is assumed to be as if you had written
```ts
const a = map["a"]; // OK per 'A' example above
```

If you write this:
```ts
declare const map: { [s: string]: string };
const { a, b }: { a: string, b: string, zz: string } = map; // ??
```
this is treated as if you had written
```ts
declare const map: { [s: string]: string };
const obj: { a: string, b: string, zz: string } = map;
const a = obj["a"], b = obj["b"];
```
Per example A, this should be OK if you interpret it as being like example 'A', but should be an error if you interpret as being like 'B'.

In conclusion, index signatures are a land of contrasts.

**Comment by MKRhere** [comment]
I don't get why example B is not okay however. Isn't it simply a narrowing a type? The runtime object might have other keys that are logically excluded with a type signature.

B would still work if it was narrowing an interface with higher number of properties than expected like so:

```ts
declare const map: { a: string, b: string, c: string, zz: string } ;
const B: { a: string, b: string, zz: string } = map; // this is OK
```

In my opinion, it's currently inconsistent behaviour, and having B behave more like A is actually consistency.

**Comment by RyanCavanaugh** [comment]
It's tempting to have index signatures appear to present *all* possible properties, but then the type `{ [k: string]: any }` becomes assignable to nearly every type, which isn't desirable at all.

**Comment by MKRhere** [comment]
> It's tempting to have index signatures appear to present _all_ possible properties

Don't they, though?

**Comment by RyanCavanaugh** [comment]
Only during property access, not during assignability checks

**Comment by MKRhere** [comment]
I don't see a reason to not have behaviour parity in both cases, but up to the team to make these difficult decisions I guess. Thanks anyway.

**Comment by fatcerberus** [comment]
In practice, a type with an index signature generally represents a very amorphous object (e.g. `Record`); you don't typically want that to be used directly in place of something with a more rigid, well-defined structure.

**Comment by fatcerberus** [comment]
It probably makes more sense if you consider the index signature to represent "every possible property but they're all optional", as this is closer to their intent in practice.
